Financial updates: LCOE calc and grouping logic

johnjasa opened the pull request:
# Financial updates: LCOE calc and grouping logic

Addressing some small financial modeling updates that came about when working on #165.
- LCOE and LCOH metrics are now computed based on lists optionally given by users
- Minor fix for grouping of financial calcs (edge case), added test

## PR Checklist

<!--Tick these boxes if they are complete, or format them as "[x]" for the markdown to render. -->
- [x] `CHANGELOG.md` has been updated to describe the changes made in this PR
- [x] Documentation
  - [x] Docstrings are up-to-date
  - [x] Related `docs/` files are up-to-date, or added when necessary
  - [x] Documentation has been rebuilt successfully
  - [x] Examples have been updated
- [x] Tests pass (If not, and this is expected, please elaborate in the tests section)
- [x] PR description thoroughly describes the new feature, bug fix, etc.

## Related issues

Resolves #175


## Impacted areas of the software


## Additional supporting information

<!--Add any other context about the problem here.-->


## Test results, if applicable

<!--
Add the results from unit tests and regression tests here along with justification for any
failing test cases.
-->


<!--
__ For NREL use __
Release checklist:
- [ ] Update the version in h2integrate/__init__.py
- [ ] Verify docs builds correctly
- [ ] Create a tag on the main branch in the NREL/H2Integrate repository and push
- [ ] Ensure the Test PyPI build is successful
- [ ] Create a release on the main branch
-->

johnjasa commented on h2integrate/core/h2integrate_model.py:
```diff
@@ -275,9 +275,16 @@ def create_financial_model(self):
+
+        for tech_name, tech_config in self.technology_config["technologies"].items():
+            if tech_name not in all_grouped_techs:
+                if 1 not in financial_groups:
```
Yes, that is correct

johnjasa replied:
Do you want that behavior to change? Maybe have a `default` group instead of `1`? Or something else?

jaredthomas68 replied:
I agree with Kaitlin here. I think the default group name should be a name that a user is not likely to set intentionally, like 'default'.

johnjasa commented on h2integrate/core/finances.py:
```diff
@@ -204,6 +207,11 @@ def compute(self, inputs, outputs):
+                last_tech_that_produces_electricity = tech
+
         # --------------------------------- Add capital and fixed items to ProFAST --------------
         for tech in self.tech_config:
```
Great point, we talked through this together and I changed the code introduced in this PR so users can directly specify which technologies are included in the metrics.

jaredthomas68 commented on tests/h2integrate/test_all_examples.py:
```diff
@@ -126,9 +126,6 @@ def test_wind_h2_opt_example(subtests):
     with subtests.test("Check LCOH"):
         assert model.prob.get_val("financials_group_1.LCOH")[0] < 4.64
```
I know this is not part of this PR per se, but It think it would be better to change the 4.64 on the right to the initial LCOH before optimization so that if the models or assumptions change, the test remains valid and useful.

johnjasa replied:
Got it, good call, I modified the case and test!

jaredthomas68 reviewed (changes requested):
Thanks for this John, looks good generally. I left a couple of small comments, the only one I see as blocking is one Kaitlin had already pointed out.

johnjasa commented:
Thanks for your reviews, @kbrunik and @jaredthomas68! Ready for re-reviews.

I've modified the default group so it's now `financial_group_default` and removed the automatic grouping based on the last tech producing electricity. Instead, users must now define exactly what technologies they want included for any LCO metric or _all_ technologies will be included.

johnjasa commented on docs/user_guide/how_to_set_up_an_analysis.ipynb:
````diff
@@ -171,6 +171,26 @@
+        "This is done by defining the list of those technologies in the `finance_parameters` section like this:\n",
+        "\n",
+        "```\n",
+        "finance_parameters:\n",
````
Fixed!
